samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -86,61 +82,34 @@ class KafkaCheckpointManager(
     val key = KafkaCheckpointLogKey.getCheckpointKey(taskName)
     val keyBytes = key.toBytes()
     val msgBytes = serde.toBytes(checkpoint)
+    val systemStream = new SystemStream(systemName, checkpointTopic)
+    val envelope = new OutgoingMessageEnvelope(systemStream, keyBytes, msgBytes)
+
     retryBackoff.run(
       loop => {
-        if (producer == null) {
-          producer = connectProducer()
+        if (systemProducer == null) {
+          systemProducer.synchronized {
+            if (systemProducer == null) {
+              systemProducer = getSystemProducer()
```

Comment:
Why do we need a double `null` check here again? All you need is a synchronized read-modify-write back to the variable right? Can we simplify this?

Reply:
this allows you to avoid getting lock in most cases.